src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserHelperClasses.cs

```diff
@@ -300,103 +300,40 @@ internal void Clear()
 
     internal static class SqlEnvChangePool
     {
-        private static int _count;
-        private static int _capacity;
-        private static SqlEnvChange[] _items;
-#if DEBUG
-        private static string[] _stacks;
-#endif 
-
-        static SqlEnvChangePool()
+        private protected struct SqlEnvChangeWrapper
         {
-            _capacity = 10;
-            _items = new SqlEnvChange[_capacity];
-            _count = 1;
-#if DEBUG
-            _stacks = new string[_capacity];
-#endif 
+            public SqlEnvChange Item;
         }
 
+        private static readonly SqlEnvChangeWrapper[] _items = new SqlEnvChangeWrapper[Environment.ProcessorCount * 2];
+
         internal static SqlEnvChange Allocate()
         {
-            SqlEnvChange retval = null;
-            lock (_items)
+            SqlEnvChangeWrapper[] items = _items;
+            for (int index = 0; index < items.Length; index++)
             {
-                while (_count > 0 && retval is null)
+                SqlEnvChange item = items[index].Item;
+                if (item != null && Interlocked.CompareExchange(ref items[index].Item, null, item) == item)
                 {
-                    int count = _count; // copy the count we think we have
-                    int newCount = count - 1; // work out the new value we want
-                                              // exchange _count for newCount only if _count is the same as our cached copy
-                    if (count == Interlocked.CompareExchange(ref _count, newCount, count))
-                    {
-                        // count is now the previous value, we're the only thread that has it, so count-1 is safe to access
-                        Interlocked.Exchange(ref retval, _items[count - 1]);
-                    }
-                    else
-                    {
-                        // otherwise the count wasn't what we expected, spin the while and try again.
-                    }
+                    return item;
                 }
             }
-            if (retval is null)
-            {
-                retval = new SqlEnvChange();
-            }
-            return retval;
+            return new SqlEnvChange();
```

Comment:
This allocated `SqlEnvChange` will not be tracked by `_items`
This seems counter intuitive to the design. 

When this item is released in the `Release` call below, it will not go back to the _items. If `_items` is not gong to be a container for all the SqlEnvChange objects, then why have it in the first place ?

Reply:
Let me know if I have missed anything

Reply:
SqlEnvChange is used in a small number of places. All allocation is done through this function. The pool is designed to be a buffer but not a block so if we have a free object hanging around then use it but if we're running fast and don't have a spare object I don't want to wait for someone to return one to the pool, just create a new one. Like allocation release is always done through the pool so if an item was new'ed rather than being sourced from the pool it will still be given back to the pool. If the pool is full the item will simply be discarded.

This behaviour means that at low frequency somewhere below the pool threshold all items are likely to be sourced from the pool which makes the operations lower memory than they otherwise would be. At high frequency the pool will be recycled quickly but some items are still going to be new'ed and dropped to GC, overall the operations will have lower GC pressure.

Reply:
I am not sure if this assumption works very well. SqlEnvChange is used in a small number of places. But then the places it is used can happen concurrently and with high through put. 

![image](https://user-images.githubusercontent.com/1623701/55826071-0540e000-5abc-11e9-9a55-d5a68ffb56f4.png)


For applications with small load, I don't care as much for this change, however the applications where the improvements is useful are web apps which will spin up a lot of connections at the same time from different threads, potentially from few connection pools. Those are the real world cases. In those cases, the number of env change token from the pool will be exhausted quickly enough. 

I don't have a quantification on how big the Env Change pool should be, but I do know that ProcessCount * 2 doesn't get close enough to the concurrent connection open requests and transaction promotions that happen in real world applications.

Reply:
Ok. then either we need a larger pool or to drop it entirely, I'm fine with either. 

In low memory cases having a very large pool would be a constant cost and possibly go mostly unused. In the case of bringing up an entire pool how many concurrent connections are likely to be in the env change code at once do you think? my instinct says it's likely to be high so a high pool would be best.

The list and array rental changes of the original RP are still beneficial to all cases. I suspect the pool is of dubious benefit without some per numbers to back it up at this point. Would you agree @AfsanehR ?

Reply:
I am leaning towards dropping this pool till we can have a more solid reasoning behind the size, and have a one size fits all model.

I don't have any ideas to improve this right now. Let's try to keep this simple (though possible less performant in some cases) and remove the pool.